File: src/advent_of_code/y2020/day15.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    int size = 1 << 25; // Enough to solve given input without reallocating
    int *seen = calloc(size, sizeof(int));
    int i = 0;
    int n;
    while (scanf("%d", &n) != EOF) {
        seen[n] = ++i;
        getchar();
    }

    int prev = i;
    while (++i <= 30000000) {
        n = i - 1 - prev;
        if (n >= size) {
            if (!(seen = realloc(seen, size * 2 * sizeof(int)))) return 1;
            memset(seen + size, 0, size * sizeof(int));
            size *= 2;
        }
        prev = seen[n] ? seen[n] : i;
        seen[n] = i;
        if (i == 2020) {
            printf("%d\n", n);
        }
    }
    printf("%d\n", n);
}

File: src/advent_of_code/y2020/day15.zig
const std = @import("std");
const root = @import("../main.zig");
const Result = root.Result;
const resizeZFill = root.support.resizeZFill;

pub const Solution = root.Solution{ .year = 2020, .day = 15, .run = run };

fn run(input: []const u8) !Result {
    const gpa = std.heap.page_allocator;
    const initSize = 1 << 25; // Enough to solve given input without reallocating
    var seen = try std.ArrayList(u32).initCapacity(gpa, initSize);
    defer seen.deinit(gpa);
    try resizeZFill(u32, &seen, gpa, initSize);

    var i: u32 = 0;
    var nums = std.mem.tokenizeAny(u8, input, ",\n");
    while (nums.next()) |num| {
        const n = try std.fmt.parseInt(u32, num, 10);
        i += 1;
        try resizeZFill(u32, &seen, gpa, n + 1);
        seen.items[n] = i;
    }

    var part1: u64 = undefined;
    var prev = i;
    var n: usize = undefined;
    while (i < 30000000) {
        n = i - prev;
        try resizeZFill(u32, &seen, gpa, n + 1);
        i += 1;
        prev = if (seen.items[n] != 0) seen.items[n] else i;
        seen.items[n] = i;
        if (i == 2020) {
            part1 = n;
        }
    }

    return Result.of(part1, n);
}
